src/common/Throttle.h

```diff
@@ -330,4 +332,84 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  class Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<int64_t> remain = { 0 }, max = { 0 };
+    Mutex lock;
+    list<Cond*> cond;
+    int64_t get(int64_t c);
+    friend class TokenBucketThrottle;
+
+  public:
+    Bucket(CephContext *cct, const std::string& n, int64_t m = 0);
+    ~Bucket();
+
+  public:
+    void set_max(int64_t m);
+    bool get(int64_t c, Context *ctx);
+    int64_t put(int64_t c);
+  };
+
+  class Blocker {
+    int64_t token_requested;
+    Context *got_token_ctx;
+
+    Blocker(int64_t _token_requested, Context *ctx)
+    : token_requested(_token_requested), got_token_ctx(ctx) {}
+
+    friend class TokenBucketThrottle;
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  int64_t m_avg;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+  list<Blocker*> m_blockers;
+  Mutex m_blockers_lock;
+
+public:
+  TokenBucketThrottle(CephContext *cct,
+		      int64_t capacity,
+		      int64_t avg,
+		      SafeTimer *timer,
+		      Mutex *timer_lock)
+  : m_cct(cct), m_throttle(m_cct, "token_bucket_throttle", capacity),
+    m_avg(avg), m_timer(timer), m_timer_lock(timer_lock),
+    m_blockers_lock("TokenBucketThrottle::blockers_lock")
+  {
+    Mutex::Locker timer_locker(*m_timer_lock);
+    add_tokens();
+  }
+  ~TokenBucketThrottle(){
+    {
+      Mutex::Locker timer_locker(*m_timer_lock);
+      cancel_tokens();
+    }
+    Mutex::Locker l(m_blockers_lock);
+    while (!m_blockers.empty()) {
+      Blocker *b = m_blockers.front();
+      m_blockers.pop_front();
+      m_blockers_lock.Unlock();
+      b->got_token_ctx->complete(-EAGAIN);
+      m_blockers_lock.Lock();
+      delete b;
+    }
+  }
+
+private:
+  void add_tokens();
```

Comment:
Nit: for clarity, perhaps these methods should be ```schedule_timer``` and ```cancel_timer``` -- and the actual ```add_tokens``` code should be separated out into its own function invoked by the timer.